Create attributes `AsTwigFilter`, `AsTwigFunction` and `AsTwigTest` to ease extension development

GromNaN opened the pull request:
One drawback to writing extensions at present is that the declaration of functions/filters/tests is not directly adjacent to the methods. It's worse for runtime extensions because they need to be in 2 different classes. See [`SerializerExtension`](https://github.com/symfony/symfony/blob/7.0/src/Symfony/Bridge/Twig/Extension/SerializerExtension.php) and [`SerializerRuntime`](https://github.com/symfony/symfony/blob/7.0/src/Symfony/Bridge/Twig/Extension/SerializerRuntime.php) as an example.

By using attributes for filters, functions and tests definition, we can make writing extensions more expressive, and use reflection to detect particular options (`needs_environment`, `needs_context`, `is_variadic`).

Example if we implemented the `formatDate` filter: https://github.com/twigphp/Twig/blob/aeeec9a5e907a79e50a6bb78979154599401726e/extra/intl-extra/IntlExtension.php#L392-L395

By using the `AsTwigFilter` attribute, it is not necessary to create the `getFilters()` method. The `needs_environment` option is detected from method signature. The name is still required as the method naming convention (camelCase) doesn't match with Twig naming convention (snake_case).

```php
use Twig\Extension\Attribute\AsTwigFilter;

class IntlExtension
{
    #[AsTwigFilter(name: 'format_date')]
    public function formatDate(Environment $env, $date, ?string $dateFormat = 'medium', string $pattern = '', $timezone = null, string $calendar = 'gregorian', string $locale = null): string
    {
        return $this->formatDateTime($env, $date, $dateFormat, 'none', $pattern, $timezone, $calendar, $locale);
    }
}
```

This approach does not totally replace the current definition of extensions, which is still necessary for advanced needs. It does, however, make for more pleasant reading and writing. 

This makes writing lazy-loaded runtime extension the easiest way to create Twig extension in Symfony: https://github.com/symfony/symfony/pull/52748

Related to https://github.com/symfony/symfony/issues/50016

Is there any need to cache the parsing of method attributes? They are only read at compile time, but that can have a performance impact during development or when using dynamic templates.

GromNaN committed:
Create attributes `AsTwigFilter`, `AsTwigFunction` and `AsTwigTest` to ease extension development

GromNaN commented on src/Extension/Extension.php:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @author Jérôme Tamarelle <jerome@tamarelle.net>
+ */
+abstract class Extension extends AbstractExtension
```
Is it better to modify the existing `AbstractExtension`, create a new abstract class or a new trait? The trait would have to be used with the `AbstractExtension` or `ExtensionInterface`.

derrabus replied:
From a DX perspective: If I used the new attributes, why would I still want to extend `AbstractExtension` directly or though this class? If we introduce those attributes, I believe we should make them work on standalone classes.

GromNaN replied:
Right. I can create a wrapper class that implements the ExtensionInterface and scan one or many object methods, passed to the constructor, to find filters/functions/tests. That would be much better. 

We need a marker to hint Symfony DI that a class contains Twig extension attributes, to avoid scanning attributes of all methods of all classes in the project.

GromNaN replied:
The marker will be `RuntimeExtensionInterface`. This interface will soon find its purpose in Twig.

derrabus replied:
If we introduce all new attributes in this PR, can't the marker be, you know, an attribute? 🥺

GromNaN replied:
I created the `AsTwigExtension` attribute as marker.

GromNaN commented on src/Extension/Extension.php:
```diff
@@ -0,0 +1,100 @@
+                    $options['is_variadic'] = $param && $param->isVariadic();
+                }
+
+                yield new TwigFilter($attribute->name ?? $method->getName(), [$this, $method->getName()], $options);
```
Should the default name be converted from camelCase to snake_case, to match the Twig convention? This makes it more difficult to find the declaration of a twig function/filter/test in PHP source code. But it does ensure that developers follow the naming convention by default.

stof replied:
Given that Twig function names must be globally unique while method names are unique only to the class, I'm not even sure we want this automatic name. I'd rather make the name mandatory in the attribute.

Thus, this name is a public API of the Twig extension, so it means that any guessing logic we introduce is then set in stone as any change to it as a BC break.

GromNaN replied:
I updated to make the function name required. A default value can always be set later.

GromNaN commented on src/Extension/Attribute/AsTwigFilter.php:
```diff
@@ -0,0 +1,29 @@
+        public ?string $name = null,
+
+        /**
+         * @var array{needs_environment?:bool, needs_context?:bool, is_variadic?:bool, is_safe?:array|null, is_safe_callback?:callable|null, pre_escape?:string|null, preserves_safety?:array|null, node_class?:class-string, deprecated?:bool|string, alternative?:string}
```
Since we can detect automatically if the function is variadic, needs the environment of the context, this option could be undocumented.
I'm not sure for the `node_class` either.

GromNaN replied:
We could flatten the options in the class properties, which would undoubtedly be easier to read and benefit from native typing.

GromNaN commented on src/Extension/Attribute/AsTwigTest.php:
```diff
@@ -0,0 +1,34 @@
+        /**
+         * @var array<int, mixed>
+         */
+        public array $arguments = [],
```
I'm not sure what `$arguments` is for.

derrabus commented on src/Extension/Attribute/AsTwigFilter.php:
```diff
@@ -0,0 +1,29 @@
+        /**
+         * @var array{is_safe?:array|null, is_safe_callback?:callable|null, pre_escape?:string|null, preserves_safety?:array|null, deprecated?:bool|string, alternative?:string}
+         */
+        public array $options = [],
```
All those options should be dedicated properties on the attribute.

GromNaN commented:
I'll rework the implementation after reading discussions on symfony/symfony#50016

stof commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,179 @@
+                    $parameters = $method->getParameters();
+                    $needsEnvironment = isset($parameters[0]) && Environment::class === $parameters[0]->getType()?->getName();
+                    $firstParam = $needsEnvironment ? 1 : 0;
+                    $needsContext = isset($parameters[$firstParam]) && 'context' === $parameters[$firstParam]->getName() && 'array' === $parameters[$firstParam]->getType()?->getName();
```
this special behavior of parameter named `$context` needs to be documented on the attributes.

GromNaN replied:
I added the doc and example in the attribute class header.

GromNaN replied:
The [Symfony `serialize` filter](https://github.com/symfony/symfony/blob/2d88c405ba0de17aebcf5df3de79b12f3fdb297a/src/Symfony/Bridge/Twig/Extension/SerializerRuntime.php#L29) accepts an `array $context` as 3rd argument. This will not conflict but this means developers could have an unexpected behavior if they don't know about this special argument handling.

GromNaN replied:
Removed this fragile detection.

GromNaN commented on src/Extension/WithLastModified.php:
```diff
@@ -0,0 +1,8 @@
+
+namespace Twig\Extension;
+
+interface WithLastModified
```
This is also a new feature required to detect when the class that defines attributes have been updated.

GromNaN replied:
Alternatively, we can make passing a class name to the ExtensionSet, that will be scanned by it. This has slightly more impact on existing classes.

GromNaN replied:
@stof I can't replace the modification date by a hash, the `Environment::isFresh` relies on a date comparison between the ExtensionSet and the cache file mtime.

https://github.com/twigphp/Twig/blob/aeeec9a5e907a79e50a6bb78979154599401726e/src/Environment.php#L416-L419

nicolas-grekas replied:
`LastModifiedAwareInterface`? What's the existing convention in Twig?

GromNaN replied:
Renamed to `ModificationAwareInterface`. The convention if to suffix with `Interface`.

GromNaN replied:
With the last change, I removed this interface. `ExtensionSet` is now aware of the extension classes that use attributes, and check their modification date as well.

GromNaN commented on src/Extension/Attribute/AsTwigExtension.php:
```diff
@@ -0,0 +1,20 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\Extension\Attribute;
```
Should it move to `Twig\Attribute` namespace? I see other libraries like Monolog and Symfony HttpKernel have the `Attribute` namespace at the root of their namespace.

GromNaN replied:
Moved.

smnandre commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,179 @@
+
+                    $name = $attribute->name;
+                    if (isset($filters[$name])) {
+                        throw new \LogicException(sprintf('Multiple definitions of the "%s" filter.', $name));
```
Could we have a unique centralized system for that ? Because not all functions/filters will come from here i guess ?

GromNaN replied:
This is a feature of Twig. You can replace an existing filter/function/test by declaring the same name in an extension.
I'm thinking about removing this check.

smnandre commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,179 @@
+                    $isVariadic = isset($parameters[$firstParam]) && end($parameters)->isVariadic();
+
+                    $tests[$name] = new TwigTest($name, [$objectOrClass, $method->getName()], [
+                        'is_variadic' => $isVariadic,
+                        'deprecated' => $attribute->deprecated,
```
\+ options `node_class` and `one_mandatory_argument` ?

 (if TwigTest keeps it [current signature](https://github.com/twigphp/Twig/blob/aeeec9a5e907a79e50a6bb78979154599401726e/src/TwigTest.php#L38))

GromNaN replied:
- `node_class` option is exclusive with the callback. It cannot used for an attribute of a method. It we extends the attributes to the class, that would be the class itself.
- I must check how arguments works with tests.

smnandre replied:
I did not realize they were 100% exclusive, thanks for the explanation !

smnandre commented on src/Extension/Attribute/AsTwigExtension.php:
```diff
@@ -0,0 +1,20 @@
+ * Identifies a class that uses PHP attributes to define filters, functions, or tests.
+ */
+#[\Attribute(\Attribute::TARGET_CLASS)]
+class AsTwigExtension
```
What is the purpose of this attribute ? I don't see (yet) how it is / will be used

GromNaN replied:
It's similar to `Twig\Extension\RuntimeExtensionInterface`, it's a marker for classes that provide extension features, to be automatically tagged by Symfony dependency injection.

stof replied:
Given that extensions must implement the ExtensionInterface anyway, there is no need for a marker attribute. The Symfony dependency injection component is also able to perform auto-configuration based on implemented interfaces (and TwigBundle already does it for the Twig ExtensionInterface

GromNaN replied:
It's more a question of DX. It's more meaningful to use an attribute than an interface as a marker as the developer will use attributes in the class. Implementing an interface without method is a deviation from the interface system.

GromNaN replied:
I've made this attribute mandatory to prevent any object from being passed by mistake. In practice, it will be mandatory anyway for Symfony autoconfiguration.

GromNaN replied:
Symfony autoconfiguration can work on method attributes, so I removed the `AsTwigExtension` attribute.

smnandre commented on tests/Extension/Fixtures/ExtensionWithAttributes.php:
```diff
@@ -0,0 +1,98 @@
+    public function variadicTest(string ...$strings)
+    {
+    }
+
```
I don't see how this class would change anything (as it's [internal to ExtensionSet](https://github.com/twigphp/Twig/blob/aeeec9a5e907a79e50a6bb78979154599401726e/src/ExtensionSet.php#L169))... but maybe just to be sure could we add a "pattern" Filter/Function/Test  ?

`pattern_*_filter`

GromNaN replied:
Test added.

GromNaN commented on doc/advanced.rst:
```diff
@@ -756,6 +756,156 @@ The ``getTests()`` method lets you add new test functions::
+
+.. note::
+
+    The ``\Twig\Extension\AttributeExtension`` can be added only once to an environment.
```
There is an exception if we try to register the extension a second time.

GromNaN reviewed (commented):
Ready for review @fabpot.

GromNaN commented:
In all honesty, I'm not totally convinced by this implementation. The fact that extensions have to be registered in 2 places (as a runtime extension and for definition in AttributeExtension) makes them difficult to use with Twig standalone, this is totally hidden for Symfony users with the TwigBundle.

I would like to add a new `Environment::add(object $extension)` method that would accept any class with attributes. Invocable classes could be used to define filters and functions.

nicolas-grekas commented:
What is a bit strange is having one single instance of AttributeExtension be responsible for registering many runtimes.
Maybe we should instead have one AttributeExtension per runtime?
Then it might be easier to use this standalone : `addExtension(new AttributeExtension(new AttributeBasedRuntime)))` (and such runtimes could have a static factory to make this even easier to create).
Note that this suggestion might be wrong as I didn't think of how runtimes should be made lazy-instantiated.

kbond commented:
For reference, this is my attempt/experiment to solve the problem of apps having to create a bunch of twig extensions for things: https://github.com/zenstruck/twig-service-bundle

GromNaN commented on src/Environment.php:
```diff
@@ -616,14 +616,17 @@ public function getRuntime(string $class)
+    /**
+     * @param ExtensionInterface|object|class-string $extension
+     */
+    public function addExtension(object|string $extension)
```
This is a breaking change.

GromNaN reviewed (commented):
> What is a bit strange is having one single instance of AttributeExtension be responsible for registering many runtimes. Maybe we should instead have one AttributeExtension per runtime? Then it might be easier to use this standalone : `addExtension(new AttributeExtension(new AttributeBasedRuntime)))` (and such runtimes could have a static factory to make this even easier to create). Note that this suggestion might be wrong as I didn't think of how runtimes should be made lazy-instantiated.

In the `ExtensionSet`, extensions are identified by their class name. The same extension class cannot be registered multiple times with distinct configuration.

I updated this PR so that extension can be registered directly using any class name or object that have the `#[AsTwigExtension]` attribute (the attribute becomes required).

```diff
    $twig = new \Twig\Environment($loader);
-    $twig->addExtension(new \Twig\Extension\AttributeExtension([
-        ProjectExtension::class,
-    ]));
+    $twig->addExtension(ProjectExtension::class);
    $twig->addRuntimeLoader(new \Twig\RuntimeLoader\FactoryLoader([
        ProjectExtension::class => function () use ($lipsumProvider) {
            return new ProjectExtension($lipsumProvider);
        },
    ]));
```

To use standalone (non-lazy-loaded):

```php
$twig->addExtension(new ProjectExtension($lipsumProvider));
```

stof commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,162 @@
+        }
+
+        // Assign all at the end to avoid inconsistent state in case of exception
+        $this->filters = $filters;
```
you should do `array_values` on those arrays instead of returning names as keys

stof commented on src/Attribute/AsTwigTest.php:
```diff
@@ -0,0 +1,50 @@
+ *
+ * If the first argument of the method has Twig\Environment type-hint, the test will receive the current environment.
+ * If the next argument of the method is named $context and has array type-hint, the test will receive the current context.
+ * The last argument of the method is the value to be tested, if any.
```
this is wrong. It is not the last argument. It is the following argument. A test can have as many arguments as you want after the value being tested, which become arguments in the Twig signature.

stof commented on src/Attribute/AsTwigFilter.php:
```diff
@@ -0,0 +1,78 @@
+        public ?array $preservesSafety = null,
+
+        /**
+         * Set to true if the filter is deprecated.
```
what is the string type about then ?

stof commented on src/Attribute/AsTwigFilter.php:
```diff
@@ -0,0 +1,78 @@
+        public ?string $preEscape = null,
+
+        /**
+         * Preserves the safety of the value that the filter is applied to.
```
this misses the doc for the type of array values

stof commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,162 @@
+                throw new \LogicException(sprintf('Extension class "%s" must have the attribute "%s" in order to use attributes', is_string($objectOrClass) ? $objectOrClass : get_debug_type($objectOrClass), AsTwigExtension::class));
+            }
+
+            foreach ($reflectionClass->getMethods() as $method) {
```
thinking about that, I'm wondering whether those `asTwigFilter`, `AsTwigFunction` and `AsTwigTest` attributes should be read by that extension at all.

An alternative solution would be to have an extension in which you inject this list of test, filter and function metadata, and letting TwigBundle perform all this attribute reading at build-time instead. 
This would even removed the need for the `AsTwigExtension` attribute entirely (which would be quite confusing as the class having this attribute is _not_ a twig extension but a twig runtime) because the autoconfiguration system is able to apply autoconfiguration based on method-level attributes already.

GromNaN replied:
I think that the logic for reading attributes should be in Twig,and the extension could receive the configuration already extracted. And Symfony TwigBundle will do the job of serializing and caching the configuration.

The `serializing` may be problematic in the future if closures are used in attributes for `safeCallback` ([RFC for PHP 8.5](https://wiki.php.net/rfc/closures_in_const_expr))

GromNaN replied:
> thinking about that, I'm wondering whether those `asTwigFilter`, `AsTwigFunction` and `AsTwigTest` attributes should be read by that extension at all.

Here is an other approach where attribute class can create the twig callable themself.
https://github.com/twigphp/Twig/blob/36a0732ee66bf405c2164282626fbdc8c23ae81a/src/Attribute/AsTwigFilter.php

stof commented on src/ExtensionSet.php:
```diff
@@ -424,9 +433,16 @@ private function initExtensions(): void
+            if ($extension instanceof ExtensionInterface) {
+                $this->initExtension($extension);
+            } else {
+                $classes[] = $extension;
```
making `$this->extensions` contain a mix of extensions and of runtimes (but not all of them) looks weird to me

stof replied:
if this is a way to solve the case of `getLastModified`, I'd rather add a way to make ExtensionSet aware of a **separate** list of runtimes needing to be checked (which would allow solving the case where we don't invalidate the cache today if you change the argument names in a runtime while it actually impacts the template compilation if you use named parameters in Twig, by making TwigBundle inject the name of _all_ runtimes in the list).

fabpot commented on doc/advanced.rst:
```diff
@@ -813,6 +813,145 @@ The ``getTests()`` method lets you add new test functions::
         // ...
     }
 
+Using PHP Attributes to define extensions
```
```suggestion
Using PHP Attributes to define Extensions
```

fabpot commented on doc/advanced.rst:
```diff
@@ -813,6 +813,145 @@ The ``getTests()`` method lets you add new test functions::
+Using PHP Attributes to define extensions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 3.9
```
```suggestion
.. versionadded:: 3.19
```

fabpot commented on doc/advanced.rst:
```diff
@@ -813,6 +813,145 @@ The ``getTests()`` method lets you add new test functions::
+
+.. versionadded:: 3.9
+
+    The attribute classes were added in Twig 3.9.
```
```suggestion
    The attribute classes were added in Twig 3.19.
```

fabpot commented on doc/advanced.rst:
```diff
@@ -813,6 +813,145 @@ The ``getTests()`` method lets you add new test functions::
+
+    The attribute classes were added in Twig 3.9.
+
+From PHP 8.0, you can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``,
```
As we now require PHP 8, the version mention can be removed:

```suggestion
You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``,
```

fabpot commented on doc/advanced.rst:
```diff
@@ -813,6 +813,145 @@ The ``getTests()`` method lets you add new test functions::
+    $twig = new \Twig\Environment($loader);
+    $twig->addExtension(ProjectExtension::class);
+
+If all the methods are static, you are done. The ``Project_Twig_Extension`` class will
```
```suggestion
If all the methods are static, you are done. The ``ProjectExtension`` class will
```

fabpot commented on src/ExtensionSet.php:
```diff
@@ -424,6 +434,7 @@ private function initExtensions(): void
         foreach ($this->extensions as $extension) {
             $this->initExtension($extension);
         }
+
```
Can be reverted

fabpot commented on tests/Extension/Fixtures/FilterWithoutValue.php:
```diff
@@ -0,0 +1,13 @@
+    public function myFilter()
+    {
+    }
+}
```
missing new line

fabpot commented on tests/Extension/Fixtures/TestWithoutValue.php:
```diff
@@ -0,0 +1,13 @@
+    public function myTest()
+    {
+    }
+}
```
missing new line

fabpot reviewed (changes requested):
Not an extensive review, just things I've spotted while reading the code quickly.

GromNaN commented:
> What is a bit strange is having one single instance of AttributeExtension be responsible for registering many runtimes. Maybe we should instead have one AttributeExtension per runtime? Then it might be easier to use this standalone : `addExtension(new AttributeExtension(new AttributeBasedRuntime)))` (and such runtimes could have a static factory to make this even easier to create). Note that this suggestion might be wrong as I didn't think of how runtimes should be made lazy-instantiated.

I refactored once again to get 1 instance of `AttributeExtension` for each class. This allows to list the encapsulated class in the `ExtensionSet::getSignature` to invalidate the cache when an extension class is removed.

Also, I removed the ability to pass an instance of an object to `AttributeExtension`, this would require proxying calls to `ExtensionSet::getExtension('MyClass')->functionCall()` to the wrapped object. This can be implemented in a follow-up PR.

The doc is up-to-date on how to use it standalone. The Symfony implementation need to be updated.

nicolas-grekas commented:
git rebase FTW ;)

GromNaN commented on src/ExtensionSet.php:
```diff
@@ -142,7 +143,11 @@ public function getLastModified(): int
     public function addExtension(ExtensionInterface $extension): void
     {
-        $class = $extension::class;
+        if ($extension instanceof AttributeExtension) {
+            $class = $extension->getClass();
+        } else {
+            $class = $extension::class;
+        }
```
This is for cache invalidation when an extension is added/removed.

Instead of relying on this class check, I can introduce an interface: `DelegatingExtensionInterface` with the `getClass` method.

GromNaN reviewed (commented):
The state of my reflection is that we would want a different implementation in Symfony TwigBundle that reads the attributes during container compilation and inject a pre-compiled instance of `AttributeExtension`.

This can be done in 2 ways:
- by serializing the `AttributeExtension` (I'm not sure for the performances),
- or by exposing an API to convert attributes to TwigCallable. Maybe creating a public method on the attribute classes `AsTwigFilter::getTwigCallable(\ReflectionFunctionAbstract $reflection): TwigFilter`

fabpot commented on doc/advanced.rst:
```diff
@@ -814,6 +814,107 @@ The ``getTests()`` method lets you add new test functions::
+Then register the ``Twig\Extension\AttributeExtension`` with the class name::
+
+    $twig = new \Twig\Environment($loader);
+    $twig->addExtension(new \Twig\Extension\AttributeExtension(ProjectExtension::class);
```
```suggestion
    $twig->addExtension(new \Twig\Extension\AttributeExtension(ProjectExtension::class));
```

fabpot commented on src/Attribute/AsTwigFilter.php:
```diff
@@ -0,0 +1,57 @@
+     * @param DeprecatedCallableInfo|null $deprecationInfo Information about the deprecation
+     */
+    public function __construct(
+        public string $name,
```
Maybe we can make it optional and use the method name if no name is provided?

GromNaN replied:
There is a conflict in name convention as method names are [camelCase](https://symfony.com/doc/current/contributing/code/standards.html#naming-conventions) and Twig function names are [snake_case](https://twig.symfony.com/doc/3.x/coding_standards.html).

This was discussed with @stof in https://github.com/twigphp/Twig/pull/3916#discussion_r1405929739.

But I agree that this is duplicate information.

fabpot replied:
Those are just conventions ;) And for single word names, that works well.

GromNaN replied:
You're right, I made the name optional again.

stof replied:
To me, the bigger argument is that Twig filter names are global, and so should be specific enough to avoid conflicts, while the method name is purely internal. For non-core extensions, filter names are generally namespaced with a prefix, but the method names in implementation don't have this prefix in the PHP code as the extension class already separates the PHP code.

GromNaN replied:
Getting the method name simplifies the creation of function/filters in the scope of an app. For shared libraries, it's still possible and encouraged to set a prefixed name.

stof replied:
Also, my previous comment (linked above) still applies: as the Twig filter name is part of the public API of packages providing an extension (it might even be their main public API if they are all about extending Twig), we won't ever be able to change the algorithm in Twig as changing the guessed name would break those packages (technically, we could by implementing both algorithm in parallel, keeping the old return value and triggering a deprecation in case the new one produces a different result, asking projects to stop relying on guessing, but this means that any change in the algorithm requires waiting for the next major version to be active).

GromNaN replied:
Keeping the name of the method is the best that can be done. Trying to apply a transformation (prefix, snake case) would make it impossible to find the name when searching in the source code. 

If I understand correctly, what you want is to force developers to think of a name? For me, that's the only good reason to make the name required.

GromNaN replied:
I reverted the default name fallback. This can be added later.

kbond commented on doc/advanced.rst:
```diff
@@ -814,6 +814,119 @@ The ``getTests()`` method lets you add new test functions::
+You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and
+``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.
+
+Create a class using this attributes::
```
```suggestion
Create a class using these attributes::
```

kbond commented on doc/advanced.rst:
```diff
@@ -814,6 +814,119 @@ The ``getTests()`` method lets you add new test functions::
+    The attribute classes were added in Twig 3.21.
+
+You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and
+``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.
```
```suggestion
You can add the ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and
``#[AsTwigTest]`` attributes to public methods of any class to define filters, functions, and tests.
```

kbond commented on doc/advanced.rst:
```diff
@@ -814,6 +814,119 @@ The ``getTests()`` method lets you add new test functions::
+        }
+
+If you want to access the current environment instance in your filter or function,
+add the ``Twig\Environment`` type to the first argument of the method::
```
Does this automatically add the `needsEnvironment` option? What about `needsContext`? Is adding `array $context` enough?

GromNaN replied:
The detection of the context is less obvious (no specific type).
And it's very rarely used: `dump` is the only that uses it in all the Twig extra, Symfony and UX packages. Leaving it explicit seems to me to be more safe.

kbond replied:
Agreed, just auto-detecting `needsEnvironment` seems best for these reasons.

GromNaN replied:
I did the detection initially: https://github.com/twigphp/Twig/blob/c5c00ed6c64ea2470f8a0dcb4c1fc18682f8fa5c/src/Extension/Extension.php

fabpot replied:
I'm confused. `needsEnvironment` is auto-detected, right?

kbond replied:
`needsEnvironment` *is* auto-detected - `needsContext` is not.

kbond reviewed (commented):
Love this!

GromNaN commented on src/Extension/AttributeExtension.php:
```diff
@@ -0,0 +1,177 @@
+ *
+ * @author Jérôme Tamarelle <jerome@tamarelle.net>
+ */
+final class AttributeExtension extends AbstractExtension
```
It would be possible to extend this class to create an extension without runtime dependency. This is interesting to simplify standalone usage only, as Symfony will abstract the runtime declaration part.

```php
class CustomExtension extends AttributeExtension
{
    public function __construct()
    {
        // This can be set automatically for child classes
        parent::__construct(self::class);
    }

    #[AsTwigFunction('foo')]
    public function foo(): string
    {
        return 'foo';
    }
}
```

Reflection for a subsequent PR.

fabpot commented:
Thank you @GromNaN.

GromNaN commented:
Documentation: [Using PHP Attributes to define Extensions](https://twig.symfony.com/doc/3.x/advanced.html#using-php-attributes-to-define-extensions)

GromNaN commented:
Documentation: [Using PHP Attributes to define Extensions](https://twig.symfony.com/doc/3.x/advanced.html#using-php-attributes-to-define-extensions)
